include/clock_control.h

```diff
@@ -23,23 +24,81 @@ extern "C" {
 /* Used to select all subsystem of a clock controller */
 #define CLOCK_CONTROL_SUBSYS_ALL	NULL
 
+/**
+ * @brief Current clock status.
+ */
+enum clock_control_status {
+	CLOCK_CONTROL_STATUS_OFF,
+	CLOCK_CONTROL_STATUS_STARTING,
+	CLOCK_CONTROL_STATUS_ON,
+	CLOCK_CONTROL_STATUS_UNKNOWN
+};
+
+typedef void (*clock_control_cb_t)(struct device *dev, void *user_data);
+
+/**
+ * @cond INTERNAL_HIDDEN
+ */
+#define Z_CLOCK_CONTROL_ASYNC_DATA_INITIALIZER(_cb, _user_data) \
+	{ \
+		.cb = cb, \
+		.user_data = _user_data \
+	}
+/**
+ * INTERNAL_HIDDEN @endcond
+ */
+
+/**
+ * Define and initialize clock_control async data.
+ *
+ * @param name		Name of the data.
+ * @param cb		Callback.
+ * @param user_data	User data
+ */
+#define CLOCK_CONTROL_ASYNC_DATA_DEFINE(name, cb, user_data) \
+	struct clock_control_async_data name = \
+		Z_CLOCK_CONTROL_ASYNC_DATA_INITIALIZER(cb, user_data)
+
+/**
+ * @brief Clock control data used for asynchronous clock enabling.
+ *
+ * @param node		Used internally for linking asynchronous requests.
+ * @param cb		Callback called when clock is started.
+ * @param user_data	User data passed as an argument in the callback.
+ */
+struct clock_control_async_data {
+	sys_snode_t node;
```

Comment:
since 'node' is read-write accessible, the async_data is stored in RAM. All parameters (deadline or timeout) should be a member in async_data struct instead of passing as additional parameter to async_on function call.

x devs -> y subsystems -> z async_data(callback, timeout/deadline/any_other_param, user_data)

Reply:
I removed deadline from it because it is possible to set async_on request without callback. Then user don't need to create structure at all. Second reason for doing that is that `async_data` may be reused by the user without any changes throughout operation it callback is the same. It can be treated then more like a handle. Only deadline changes.

Reply:
_async_data_ being used in slist each structure instance has to be unique or you will need a mutex in each structure to avoid memory corruption. Example, if _async_on_ is called n times with same _async_data_

Reply:
yes, each caller needs unique _async_data_ and error is returned if it is detected that requested _async_data_ is already in the list.